metrics/density/fast_footprint.sh

```diff
@@ -0,0 +1,556 @@
+	((mapped*=1024))
+
+	# Cached
+	local meminfo_cached=$(grep "^Cached:" /proc/meminfo | awk '{print $2}')
+	((meminfo_cached*=1024))
+
+	local json="$(cat << EOF
+		"system": {
+			"avail": $avail,
+			"avail_decr": $avail_decr,
+			"cached": $cached,
+			"smem_free": $smem_free,
+			"free_decr": $free_decr,
+			"anon": $anon,
+			"mapped": $mapped,
+			"meminfo_cached": $meminfo_cached,
+			"Units": "KB"
+		}
+EOF
+)"
+
+	metrics_json_add_array_fragment "$json"
+}
+
+function grab_stats() {
+	# If configured, dump the caches so we get a more stable
+	# view of what our static footprint really is
+	if [[ DUMP_CACHES ]] ; then
+		dump_caches
+	fi
+
+	# user space data
+		# USS taken by CC components
+	grab_vm_uss
+		# PSS taken all userspace
+	grab_vm_pss
+		# PSS taken all userspace
+	grab_all_pss
+		# PSS taken by dockerd
+	grab_dockerd_pss
+		# user as reported by smem
+	grab_user_smem
+
+	# System overview data
+		# System free and cached
+	grab_system
+
+	# kernel data
+		# The 'total kernel space taken' we can work out as:
+		# ktotal = ((free-avail)-user)
+		# So, we don't grab that number from smem, as that is what it does
+		# internally anyhow.
+		# Still try to grab any finer kernel details that we can though
+
+		# totals from slabinfo
+	grab_slab
+
+	metrics_json_close_array_element
+}
+
+function check_limits() {
+	mem_free=$(get_memfree)
+	if ((mem_free <= MIN_MEMORY_FREE)); then
+		echo 1
+		return
+	fi
+
+	mem_consumed=$((base_mem_avail-mem_free))
+	if ((mem_consumed >= MAX_MEMORY_CONSUMED)); then
+		echo 1
+		return
+	fi
+
+	echo 0
+}
+
+launch_containers() {
+	local parloops leftovers
+
+	(( parloops=${NUM_CONTAINERS}/${PARALLELISM} ))
+	(( leftovers=${NUM_CONTAINERS} - (${parloops}*${PARALLELISM}) ))
+
+	echo "Launching ${parloops}x${PARALLELISM} containers + ${leftovers} etras"
+
+	local iter n
+	for iter in $(seq 1 $parloops); do
+		echo "Launch iteration ${iter}"
+		for n in $(seq 1 $PARALLELISM); do
+			docker run --rm -d --runtime=$RUNTIME $PAYLOAD_RUNTIME_ARGS $PAYLOAD $PAYLOAD_ARGS &
+		done
+
+		if [[ $PAYLOAD_SLEEP ]]; then
+			sleep $PAYLOAD_SLEEP
+		fi
+
+		# check if we have hit one of our limits and need to wrap up the tests
+		if (($(check_limits))); then
+			echo "Ran out of resources, check_limits failed"
+			return
+		fi
+	done
+
+	for n in $(seq 1 $leftovers); do
+		docker run --rm -d --runtime=$RUNTIME $PAYLOAD_RUNTIME_ARGS $PAYLOAD $PAYLOAD_ARGS &
+	done
+}
+
+wait_containers() {
+	local t numcontainers
+	# nap 3s between checks
+	local step=3
+
+	for ((t=0; t<${DOCKER_POLL_TIMEOUT}; t+=step)); do
+
+		numcontainers=$(docker ps -qa | wc -l)
+
+		if (( numcontainers >=  ${NUM_CONTAINERS} )); then
+			echo "All containers now launched (${t}s)"
+				return
+		else
+			echo "Waiting for containers to launch (${numcontainers} at ${t}s)"
+		fi
+		sleep 3
+	done
+
+	echo "Timed out waiting for containers to launch (${t}s)"
+	cleanup
+	die "Timed out waiting for containers to launch (${t}s)"
```

Comment:
As we have this `die`, do we need the `echo` above with the same message?

Reply:
I put the two in as the cleanup can take some time (depending on how many containers one has run). When running interactively I wanted some indication that we'd died/timed out on screen before the final death.